[5.9] 'make:' command custom class extends

pippinsmith opened the pull request:
### Inspiration
Inspiration was drawn from [Tweet A by Taylor](https://twitter.com/taylorotwell/status/1144430661572407296),
[Tweet B in thread by Taylor](https://twitter.com/taylorotwell/status/1144441642390380544)
and [Reply in thread by Victor Bolshov](https://twitter.com/crocodile2u/status/1144514813072199680)

## Benefit
The benefit of this PR is allowing the user to define their own baseclasses for any make command instead of having to overwrite the default imported Illuminate/Laravel ones for every created file, which gets cumbersome in large projects

This PR currently allows to define your own Request baseclass in a new console.php config file as this is a class I personally have extended in company projects, and is one of the few make commands that only has a single stub where this functionality makes sense.

## Drawbacks
As mentioned in [tweet B](https://twitter.com/taylorotwell/status/1144441642390380544)  there are multiple stubs for some commands making this a complex change needing extension of various MakeCommand classes to handle switching stubs.

I'm willing to invest the time to extend this PR to all other make: commands where either there are multiple stubs depending on command input switches, or where it makes less sense to extend a base class.

## Alternative
An alternative approach could be a new command input to pass the parent class but that would defeat the purpose of having it set as a default (it could still be added to use as an override for very large projects) and it would conflict in the case of make:controller since it already has a --parent input.

### Reason for draft
However I first put this PR in draft mode to allow for input if this would be a desirable way of handling base classes (as opposed to the magic detection proposed in Taylor's tweet) and if I should continue this PR to a full implementation.

### Targeted version
It is targeted at 5.9 because, while technically it can work in 5.8 without breaking backwards compatibility, it needs a new config file (the existing config files didn't seem appropriate for this behavior)
and I don't know if adding a new config file can be done in a minor version update like that (since that would go in laravel/laravel).

### Closings 
If this all seems too much for a core update than that's fine with me, though to me it feels like putting this functionality in a 3rd party package is not ideal either, as to my knowledge it's not directly possible to override an existing make: command.

retry of #29012 because I failed to remove the merge commit and didn't start from the correct branch

Thanks in advance for your consideration and input on this feature

CyrilMazur committed:
Store the Mailgun message ID in the header X-Mailgun-Message-ID after the message has been sent

taylorotwell committed:
Merge pull request #28994 from CyrilMazur/5.8-mailgun-message-id

[5.8] Storing Mailgun Message-ID in the headers after sending

antonkomarev committed:
Add singletonIf method to the Container

pippinsmith committed:
updated GeneratorCommand::buildClass to be able to use your own Baseclass for any "make:" command

currently only implemented make:request

pippinsmith committed:
updated GeneratorCommand::buildClass to be able to use your own Baseclass for any "make:" command

currently only implemented make:request

pepijn-de-wachter commented on src/Illuminate/Console/GeneratorCommand.php:
```diff
@@ -205,6 +205,59 @@ protected function replaceClass($stub, $name)
+        $replace = [];
+
+        if ($this->getNamespace($parent) === $this->getNamespace($name) || $parent === '') {
+            $replace['use NamespacedDummyParent;'.PHP_EOL] = '';
```
inspired by ControllerMakeCommand Line 96 `$replace["use {$controllerNamespace}\Controller;\n"] = '';`
I noticed that \n doesn't work on Windows and there are no tests written for this behavior to detect it, on Windows replacing it with \r\n works, but that's why PHP_EOL exists and it works (did not test on other operating systems).

taylorotwell committed:
Merge pull request #29014 from cybercog/add/singleton-if-method-to-container

[5.9] Add singletonIf method to Container

driesvints committed:
Only test on Redis 4.3.0 for now

PhpRedis 5.0.0 was released with some breaking changes which we'll need to fix first before we can run builds on it.

driesvints committed:
Merge branch '5.8'

driesvints commented:
@pepijn-de-wachter can you rebase with master branch? Build should pass after that.

pippinsmith committed:
rebase to master

driesvints commented:
Looks like something went wrong again with Git. It's best that you try to submit again with a clean branch. 

If you need help with rebasing, here's a good article on the subject by Atlassian: https://www.atlassian.com/git/tutorials/merging-vs-rebasing
